File: JavaScript/948-Bag of Tokens.js
/**
 * @param {number[]} tokens
 * @param {number} power
 * @return {number}
 */
var bagOfTokensScore = function(tokens, power) {
    if (!tokens.length) return 0;
    let left = 0, right = tokens.length-1, res = 0;
    tokens.sort(function(a,b){
        return a-b;
    });
    
    while (left <= right && power >= tokens[left]){
        power -= tokens[left];
        res++;left++;
    }
    
    if (res > 0){
        while (left < right){
            power += tokens[right];
            res--; right--;
            while (left <= right && power>=tokens[left]){
                power -= tokens[left];
                res++;left++;
            }
        }
    }
    
    return res;
};

File: C++/948-Bag of Tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(), tokens.end());
        int res = 0, score = 0;
        int left=0, right=tokens.size()-1;
        while (left <= right) {
            if (power < tokens[left] && score > 0) {
                score--;
                power += tokens[right];
                right--;
            } else if (power >= tokens[left]) {
                score++;
                power -= tokens[left];
                left++;
                res = max(res, score);
            } else {
                break;
            }
        }
        return res;
    }
};